ssl/statem/extensions_srvr.c

```diff
@@ -803,7 +803,7 @@ int tls_parse_ctos_cookie(SSL *s, PACKET *pkt, unsigned int context, X509 *x,
 
     /* We tolerate a cookie age of up to 10 minutes (= 60 * 10 seconds) */
     now = (unsigned long)time(NULL);
-    if (tm > now || (now - tm) > 600) {
+    if ((uint32_t)(now - tm) > 600) {
```

Comment:
Internally, we should really be using `time_t` and not casting to `unsigned long` or (uint32_t) unless it's being returned from a public API.

>consider https://github.com/openssl/openssl/pull/17658 where the internal type was changed from long to time_t in 3.0 while the public API kept long. That is just plain broken.

Yes and no. The public API is certainly broken; the internal use of `time_t` isn't. I don't think there was a will to change public APIs to use time_t in the 3.0 release, due to backwards compatibility concerns. Although we *did* change the options to use uint64_t...

Reply:
Examples such as #8687 updated the types for `SSL_SESSION` to be `time_t`; as an example of how to deal with this.

Reply:
> Why is this cast needed? `now - tm` is already evaluated in `unsigned long`, and I don't see that casting the result of the subtraction is needed in order to compare against 600.

although tm is of type `unsigned long` it is only a 4 byte value, see here:
https://github.com/openssl/openssl/blob/d2d2401aed7ff45f4c013201944e1218dce12da7/ssl/statem/extensions_srvr.c#L796
and here:
https://github.com/openssl/openssl/blob/d2d2401aed7ff45f4c013201944e1218dce12da7/include/internal/packet.h#L217-L228
Well, it will not break in 2038, but in 2106, but it will eventually break.

Reply:
> Examples such as #8687 updated the types for `SSL_SESSION` to be `time_t`; as an example of how to deal with this.

This would only be okay if the API changes as well.  If the API cannot change, this simply breaks the API.
When the API cannot change, we should make the internal time representation a cyclic time model.

Reply:
> although tm is of type unsigned long it is only a 4 byte value, see here:

Yes, I saw that before I commented.

Please spell out a plausible scenario where the result of evaluating `(uint32_t)(now - tm) > 600` and the result of evaluating `(now - tm) > 600` will be different.  We're measuring the delay between when we issued the cookie and when we're processing the copy sent back to us.  If our clock went backwards, both expressions will produce a large value for the LHS since the nominally negative value "wraps around" in unsigned arithmetic, and it seems physically implausible for us to wait more than 2^32 seconds between issuing and receiving the cookie.

I don't see a part of this particular codeline that is using absolute time and thus subject to 2038 or 2106 issues -- I only see unsigned arithmetic that's comparing time differences against a duration.

Reply:
Oh, I really thought that would be obvious:
let `unsigned long` be 64 byte long, and `time_t` be 64 byte long as well,
and the year after 2106.
Then (now - tm) will always be > 2^32.
QED.

Reply:
>Well, it will not break in 2038, but in 2106, but it will eventually break.

This is a flawed assumption. 32-bit `time_t` is signed, it does not support any date beyond 2038, nor before 1902. you cannot cast a signed `time_t` to an unsigned type and assume it will work beyond 2038 (as you are ignoring dates before 1970). Most software in a 32-bit `time_t` system will be completely broken in 2038.

Reply:
No, I am of course talking about systems using 64 bit time_t, and this silly code will break, because long is a compiler dependent thing.